cpu/sam0_common/sam0_sdhc/sdhc.c

```diff
@@ -0,0 +1,1062 @@
+    SDHC_DEV->CCR.reg &= ~(SDHC_CCR_USDCLKFSEL_Msk | SDHC_CCR_SDCLKFSEL_Msk);
+    SDHC_DEV->CCR.reg |= SDHC_CCR_SDCLKFSEL(div) | SDHC_CCR_USDCLKFSEL(div >> 8);
+    SDHC_DEV->CCR.reg |= SDHC_CCR_INTCLKEN;  /* enable internal clock       */
+    while (!SDHC_DEV->CCR.bit.INTCLKS) {}    /* wait for clock to be stable */
+    SDHC_DEV->CCR.reg |= SDHC_CCR_SDCLKEN;   /* enable clock to card        */
+}
+
+/**
+ * _set_hc selects either one or four bit mode, low/high speed and clock
+ *
+ * bitwidth is SDHC_HC1R_DW_1BIT_Val or SDHC_HC1R_DW_4BIT_Val
+ * speed is SDHC_HC1R_HSEN_NORMAL_Val or SDHC_HC1R_HSEN_HIGH_Val
+ */
+static void _set_hc(sdhc_state_t *state)
+{
+    if (state->high_speed) {
+        SDHC_DEV->HC1R.reg |= SDHC_HC1R_HSEN;
+    }
+    else {
+        SDHC_DEV->HC1R.reg &= ~SDHC_HC1R_HSEN;
+    }
+    if (!SDHC_DEV->HC2R.bit.PVALEN) {       /* PVALEN is probably always low */
+        _set_speed(state, state->clock);
+    }
+    if (state->bus_width == 4) {
+        /* set four bit mode */
+        SDHC_DEV->HC1R.reg |= SDHC_HC1R_DW;
+    }
+    else {
+        /* set one bit mode */
+        SDHC_DEV->HC1R.reg &= ~SDHC_HC1R_DW;
+    }
+}
+
+/**
+ * @brief Ask to all cards to send their operations conditions (MCI only).
+ * - ACMD41 sends operation condition command.
+ * - ACMD41 reads OCR
+ *
+ * @param state pointer to sdhc
+ *
+ * @return true if success, otherwise false
+ */
+static bool _test_voltage(sdhc_state_t *state, bool f8)
+{
+    uint32_t arg;
+    uint32_t retry = 2100;
+    uint32_t response;
+
+    /*
+     * Timeout 1s = 400KHz / ((6+6+6+6)*8) cycles = 2100 retry
+     * 6 = cmd byte size
+     * 6 = response byte size
+     * 6 = cmd byte size
+     * 6 = response byte size
+     */
+    retry = 2100;
+    do {
+        /* CMD55 - Indicate to the card that the next command is an
+         * application specific command rather than a standard command.*/
+        if (!sdhc_send_cmd(state, SDMMC_CMD55_APP_CMD, 0)) {
+            return false;
+        }
+
+        /* (ACMD41) Sends host OCR register */
+        arg = OCR_VDD_27_28 | OCR_VDD_28_29 | OCR_VDD_29_30 | OCR_VDD_30_31 |
+              OCR_VDD_31_32 | OCR_VDD_32_33;
+        if (f8) { /* if not legacy card */
+            arg |= SD_ACMD41_HCS;
+        }
+
+        /* Check response */
+        if (!sdhc_send_cmd(state, SD_MCI_ACMD41_SD_SEND_OP_COND, arg)) {
+            return false;
+        }
+        response = SDHC_DEV->RR[0].reg;
+        if (response & OCR_POWER_UP_BUSY) {
+            /* Card is ready */
+            if ((response & OCR_CCS) != 0) {
+                state->type |= CARD_TYPE_HC;
+            }
+            break;
+        }
+    } while (retry--);
+
+    return retry;
+}
+
+/**
+ * \brief CMD9: Addressed card sends its card-specific
+ * data (CSD) on the CMD line mci.
+ *
+ * \return true if success, otherwise false
+ */
+static bool _test_capacity(sdhc_state_t *state)
+{
+
+    uint32_t transfer_speed;
+    uint8_t csd[CSD_REG_BSIZE];
```

Comment:
```suggestion
    /* same alignment as uint32_t to be able to quickly convert endianess */
    alignas(uint32_t) uint8_t csd[CSD_REG_BSIZE];
```

See line 505.

Requires `#include <stdalign.h>`

Reply:
I'm pretty sure arrays are always word-aligned.

Reply:
When would this not be word-aligned?